pkg/sql/logictest/tests/local/local_test: TestLogic_synthetic_privileges failed

pkg/sql/logictest/tests/local/local_test.TestLogic_synthetic_privileges [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20869069?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20869069?buildTab=artifacts#/) on master @ [7eb4054a41ad06ee49f027e41c34a26e8a9c59fe](https://github.com/cockroachdb/cockroach/commits/7eb4054a41ad06ee49f027e41c34a26e8a9c59fe):


```
=== RUN   TestLogic_synthetic_privileges
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/c6737afbd6b64f789c37699347452991/logTestLogic_synthetic_privileges587234414
    test_log_scope.go:82: use -show-logs to present logs inline
[14:46:35] setting distsql_workmem='11850B';
[14:46:37] --- done: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1205/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/synthetic_privileges with config local: 87 tests, 2 failures
[14:46:38] --- total progress: 87 statements
--- total: 87 tests, 2 failures
    logic.go:4770: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/c6737afbd6b64f789c37699347452991/logTestLogic_synthetic_privileges587234414
--- FAIL: TestLogic_synthetic_privileges (3.65s)
=== RUN   TestLogic_synthetic_privileges/view_system_table
    logic.go:3217: 
         
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1205/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/synthetic_privileges:409: SELECT username, "hashedPassword" FROM system.users WHERE username = 'testuser'
        expected success, but found
        (42501) user testuser does not have SELECT privilege on relation users
        errors.go:527: in NewInsufficientPrivilegeOnDescriptorError()
    logic.go:2398: 
         /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1205/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/logictest/tests/local/local_test_/local_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/logictest/testdata/logic_test/synthetic_privileges:414: too many errors encountered, skipping the rest of the input
--- FAIL: TestLogic_synthetic_privileges/view_system_table (0.03s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_synthetic_privileges.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57944

--- a/pkg/sql/catalog/descs/collection.go
+++ b/pkg/sql/catalog/descs/collection.go
@@ -1590,7 +1590,7 @@ func (tc *Collection) LockDescriptorWithLease(
 		return uint64(vo.Desc().GetVersion()), nil
 	}
 	// Otherwise, we should be able to lease the relevant object out.
-	desc, shouldReadFromStore, err := tc.leased.getByID(ctx, txn, id)
+	desc, shouldReadFromStore, err := tc.leased.getByID(ctx, txn, id, false /* withoutLockedTimestamp */)
 	if err != nil {
 		return 0, err
 	}

--- a/pkg/sql/catalog/descs/descriptor.go
+++ b/pkg/sql/catalog/descs/descriptor.go
@@ -441,7 +441,7 @@ func (q *byIDLookupContext) lookupLeased(
 	if q.tc.cr.IsDescIDKnownToNotExist(id, q.flags.descFilters.maybeParentID) {
 		return nil, catalog.NoValidation, catalog.NewDescriptorNotFoundError(id)
 	}
-	desc, shouldReadFromStore, err := q.tc.leased.getByID(q.ctx, q.tc.deadlineHolder(q.txn), id)
+	desc, shouldReadFromStore, err := q.tc.leased.getByID(q.ctx, q.tc.deadlineHolder(q.txn), id, q.flags.descFilters.withoutLockedTimestamp)
 	if err != nil || shouldReadFromStore {
 		// Leasing does not support leasing adding descriptors, and in certain contexts,
 		// we may want them leased. So, we will fallback to the KV based reads if requested.
@@ -654,7 +654,7 @@ func (tc *Collection) getNonVirtualDescriptorID(
 			return continueLookups, descpb.InvalidID, nil
 		}
 		ld, shouldReadFromStore, err := tc.leased.getByName(
-			ctx, tc.deadlineHolder(txn), parentID, parentSchemaID, name,
+			ctx, tc.deadlineHolder(txn), parentID, parentSchemaID, name, flags.descFilters.withoutLockedTimestamp,
 		)
 		if err != nil {
 			return haltLookups, descpb.InvalidID, err

--- a/pkg/sql/catalog/descs/getters.go
+++ b/pkg/sql/catalog/descs/getters.go
@@ -482,6 +482,10 @@ type descFilters struct {
 	// maybeParentID specifies, when set, that the looked-up descriptor
 	// should have the same parent ID, when set.
 	maybeParentID descpb.ID
+	// withoutLockedTimestamp, specifies that leased descriptors should
+	// be resolved at the current timestamp. This is to ensure caches
+	// are up to date.
+	withoutLockedTimestamp bool
 }
 
 func defaultUnleasedFlags() (f getterFlags) {
@@ -570,6 +574,13 @@ func (b ByIDGetterBuilder) WithoutOtherParent(parentID catid.DescID) ByIDGetterB
 	return b
 }
 
+// WithoutLockedTimestamp configures the ByIDGetterBuilder to ensure
+// that leased descriptors are looked up at the read timestamp.
+func (b ByIDGetterBuilder) WithoutLockedTimestamp() ByIDGetterBuilder {
+	b.flags.descFilters.withoutLockedTimestamp = true
+	return b
+}
+
 // Get builds a ByIDGetter.
 func (b ByIDGetterBuilder) Get() ByIDGetter {
 	if b.flags.isMutable {
@@ -631,6 +642,13 @@ type ByNameGetterBuilder getterBase
 // of offline descriptors.
 func (b ByNameGetterBuilder) WithOffline() ByNameGetterBuilder {
 	b.flags.descFilters.withoutOffline = false
+	return b
+}
+
+// WithoutLockedTimestamp configures the ByNameGetterBuilder to ensure
+// that leased descriptors are looked up at the read timestamp.
+func (b ByNameGetterBuilder) WithoutLockedTimestamp() ByNameGetterBuilder {
+	b.flags.descFilters.withoutLockedTimestamp = true
 	return b
 }
 

--- a/pkg/sql/catalog/descs/leased_descriptors.go
+++ b/pkg/sql/catalog/descs/leased_descriptors.go
@@ -259,6 +259,7 @@ func (ld *leasedDescriptors) getByName(
 	parentID descpb.ID,
 	parentSchemaID descpb.ID,
 	name string,
+	withoutLockedTimestamp bool,
 ) (desc catalog.Descriptor, shouldReadFromStore bool, err error) {
 	// First, look to see if we already have the descriptor.
 	// This ensures that, once a SQL transaction resolved name N to id X, it will
@@ -273,7 +274,11 @@ func (ld *leasedDescriptors) getByName(
 		return desc, false, nil
 	}
 	ld.maybeInitReadTimestamp(ctx, txn)
-	ldesc, err := ld.lm.AcquireByName(ctx, ld.leaseTimestamp, parentID, parentSchemaID, name)
+	timestamp := ld.leaseTimestamp
+	if withoutLockedTimestamp {
+		timestamp = lease.TimestampToReadTimestamp(txn.ReadTimestamp())
+	}
+	ldesc, err := ld.lm.AcquireByName(ctx, timestamp, parentID, parentSchemaID, name)
 	const setTxnDeadline = true
 	return ld.getResult(ctx, txn, setTxnDeadline, ldesc, err)
 }
@@ -281,14 +286,18 @@ func (ld *leasedDescriptors) getByName(
 // getByID return a leased descriptor valid for the transaction,
 // acquiring one if necessary.
 func (ld *leasedDescriptors) getByID(
-	ctx context.Context, txn deadlineHolder, id descpb.ID,
+	ctx context.Context, txn deadlineHolder, id descpb.ID, withoutLockedTimestamp bool,
 ) (_ catalog.Descriptor, shouldReadFromStore bool, _ error) {
 	// First, look to see if we already have the table in the shared cache.
 	if cached := ld.getCachedByID(ctx, id); cached != nil {
 		return cached, false, nil
 	}
 	ld.maybeInitReadTimestamp(ctx, txn)
-	desc, err := ld.lm.Acquire(ctx, ld.leaseTimestamp, id)
+	timestamp := ld.leaseTimestamp
+	if withoutLockedTimestamp {
+		timestamp = lease.TimestampToReadTimestamp(txn.ReadTimestamp())
+	}
+	desc, err := ld.lm.Acquire(ctx, timestamp, id)
 	const setTxnDeadline = false
 	return ld.getResult(ctx, txn, setTxnDeadline, desc, err)
 }

--- a/pkg/sql/catalog/descs/table.go
+++ b/pkg/sql/catalog/descs/table.go
@@ -19,7 +19,7 @@ import (
 func (tc *Collection) GetLeasedImmutableTableByID(
 	ctx context.Context, txn *kv.Txn, tableID descpb.ID,
 ) (catalog.TableDescriptor, error) {
-	desc, _, err := tc.leased.getByID(ctx, tc.deadlineHolder(txn), tableID)
+	desc, _, err := tc.leased.getByID(ctx, tc.deadlineHolder(txn), tableID, false /* withoutLockedTimestamp */)
 	if err != nil || desc == nil {
 		return nil, err
 	}

--- a/pkg/sql/rolemembershipcache/cache.go
+++ b/pkg/sql/rolemembershipcache/cache.go
@@ -69,7 +69,7 @@ func NewMembershipCache(
 func (m *MembershipCache) RunAtCacheReadTS(
 	ctx context.Context, db descs.DB, txn descs.Txn, f func(context.Context, descs.Txn) error,
 ) error {
-	tableDesc, err := txn.Descriptors().ByIDWithLeased(txn.KV()).Get().Table(ctx, keys.RoleMembersTableID)
+	tableDesc, err := txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.RoleMembersTableID)
 	if err != nil {
 		return err
 	}
@@ -144,7 +144,7 @@ func (m *MembershipCache) GetRolesForMember(
 	}
 
 	// Lookup table versions.
-	roleMembersTableDesc, err := txn.Descriptors().ByIDWithLeased(txn.KV()).Get().Table(ctx, keys.RoleMembersTableID)
+	roleMembersTableDesc, err := txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.RoleMembersTableID)
 	if err != nil {
 		return nil, err
 	}
@@ -155,12 +155,12 @@ func (m *MembershipCache) GetRolesForMember(
 	}
 
 	if txn.SessionData().AllowRoleMembershipsToChangeDuringTransaction {
-		systemUsersTableDesc, err := txn.Descriptors().ByIDWithLeased(txn.KV()).Get().Table(ctx, keys.UsersTableID)
+		systemUsersTableDesc, err := txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.UsersTableID)
 		if err != nil {
 			return nil, err
 		}
 
-		roleOptionsTableDesc, err := txn.Descriptors().ByIDWithLeased(txn.KV()).Get().Table(ctx, keys.RoleOptionsTableID)
+		roleOptionsTableDesc, err := txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.RoleOptionsTableID)
 		if err != nil {
 			return nil, err
 		}

--- a/pkg/sql/sessioninit/cache.go
+++ b/pkg/sql/sessioninit/cache.go
@@ -138,11 +138,11 @@ func (a *Cache) GetAuthInfo(
 		if err := txn.Descriptors().MaybeSetReplicationSafeTS(ctx, txn.KV()); err != nil {
 			return err
 		}
-		usersTableDesc, err = txn.Descriptors().ByIDWithLeased(txn.KV()).Get().Table(ctx, keys.UsersTableID)
+		usersTableDesc, err = txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.UsersTableID)
 		if err != nil {
 			return err
 		}
-		roleOptionsTableDesc, err = txn.Descriptors().ByIDWithLeased(txn.KV()).Get().Table(ctx, keys.RoleOptionsTableID)
+		roleOptionsTableDesc, err = txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.RoleOptionsTableID)
 		return err
 	})
 	if err != nil {
@@ -306,7 +306,7 @@ func (a *Cache) GetDefaultSettings(
 	err = db.DescsTxn(ctx, func(
 		ctx context.Context, txn descs.Txn,
 	) error {
-		dbRoleSettingsTableDesc, err = txn.Descriptors().ByIDWithLeased(txn.KV()).Get().Table(ctx, keys.DatabaseRoleSettingsTableID)
+		dbRoleSettingsTableDesc, err = txn.Descriptors().ByIDWithLeased(txn.KV()).WithoutLockedTimestamp().Get().Table(ctx, keys.DatabaseRoleSettingsTableID)
 		if err != nil {
 			return err
 		}

--- a/pkg/sql/syntheticprivilegecache/cache.go
+++ b/pkg/sql/syntheticprivilegecache/cache.go
@@ -66,7 +66,7 @@ func New(
 func (c *Cache) Get(
 	ctx context.Context, txn isql.Txn, col *descs.Collection, spo syntheticprivilege.Object,
 ) (*catpb.PrivilegeDescriptor, error) {
-	_, desc, err := descs.PrefixAndTable(ctx, col.ByNameWithLeased(txn.KV()).Get(), syntheticprivilege.SystemPrivilegesTableName)
+	_, desc, err := descs.PrefixAndTable(ctx, col.ByNameWithLeased(txn.KV()).WithoutLockedTimestamp().Get(), syntheticprivilege.SystemPrivilegesTableName)
 	if err != nil {
 		return nil, err
 	}

